File: src/Reverse_Linked_List.py
# -*- coding: utf-8 -*-


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if head is None:
            return None
        l, node = [], head
        while node:
            l.append(node)
            node = node.next
        for i in range(len(l) - 1, 0, -1):
            l[i].next = l[i - 1]
        l[0].next = None
        return l[-1]

File: src/algorithms/cpp/Reverse_Linked_List.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <stack>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
 
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        stack <ListNode *> st;
        ListNode *node = head, *new_head = new ListNode(0), *pre = new_head;
        while (node) {
            st.push(node);
            node = node->next;
        }
        while (!st.empty()) {
            ListNode *top = st.top();
            pre->next = top;
            pre = top;
            pre->next = NULL;
            st.pop();
        }
        return new_head->next;
    }
};

int main() {
    return 0;
}
